jlog.c

```diff
@@ -2190,13 +2234,134 @@ int jlog_ctx_read_message(jlog_ctx *ctx, const jlog_id *id, jlog_message *m) {
+  }
+
+  for (i=0; i < count; i++) {
+    msg = &m[i];
+    switch(read_method) {
+      case JLOG_READ_METHOD_MMAP:
+        memcpy(&msg->aligned_header, ((u_int8_t *)ctx->mmap_base) + data_off_iter, hdr_size);
+        break;
+      case JLOG_READ_METHOD_PREAD:
+        if (!jlog_file_pread(ctx->data, &msg->aligned_header, hdr_size, data_off_iter)) {
+          SYS_FAIL(JLOG_ERR_IDX_READ);
+        }
+        break;
+      default:
+        SYS_FAIL(JLOG_ERR_NOT_SUPPORTED);
+        break;
+    }
+    msg->header = &msg->aligned_header;
+    uncompressed_size += msg->header->mlen;
+    data_off_iter += (hdr_size + msg->header->compressed_len);
+  }
+  data_off_iter = data_off;
+  if (ctx->mess_data_size < uncompressed_size) {
+    ctx->mess_data = realloc(ctx->mess_data, uncompressed_size + 1);
+    ctx->mess_data_size = uncompressed_size;
+  }
+  char *uncompressed_data_ptr = ctx->mess_data;
+  for (i=0; i < count; i++) {
+    msg = &m[i];
+    switch(read_method) {
+      case JLOG_READ_METHOD_MMAP:
+        jlog_decompress((((char *)ctx->mmap_base) + data_off_iter + hdr_size),
+                        msg->header->compressed_len, uncompressed_data_ptr, msg->header->mlen);
+        break;
+      case JLOG_READ_METHOD_PREAD:
+        if (ctx->compressed_data_buffer_len < msg->aligned_header.compressed_len) {
+          ctx->compressed_data_buffer_len = msg->aligned_header.compressed_len * 2;
+          ctx->compressed_data_buffer = realloc(ctx->compressed_data_buffer, ctx->compressed_data_buffer_len);
+        }
+        if (!jlog_file_pread(ctx->data, ctx->compressed_data_buffer, msg->header->compressed_len, data_off_iter + hdr_size)) {
+          SYS_FAIL(JLOG_ERR_IDX_READ);
+        }
+        jlog_decompress((char *)ctx->compressed_data_buffer,
+                        msg->header->compressed_len, uncompressed_data_ptr, msg->header->mlen);
+        break;
+      default:
+        SYS_FAIL(JLOG_ERR_NOT_SUPPORTED);
+        break;
+    }
+    msg->mess_len = msg->header->mlen;
+    msg->mess = uncompressed_data_ptr;
+    data_off_iter += (hdr_size + msg->header->compressed_len);
+    uncompressed_data_ptr += msg->header->mlen;
+  }
+ finish:
+  if(ctx->last_error == JLOG_ERR_SUCCESS) {
+    return count;
+  }
+  return -1;
+}
+
+static int __jlog_ctx_bulk_pread_messages_uncompressed(jlog_ctx *ctx, const jlog_id *id, const int count,
+                                                       jlog_message *m, u_int64_t data_off) {
+  assert(!IS_COMPRESS_MAGIC(ctx));
+
+  int i = 0;
+  uint64_t total_size = 0;
+  const size_t hdr_size = sizeof(jlog_message_header);
+  jlog_message *msg = NULL;
+  u_int64_t data_off_iter = data_off;
+
+  for (i=0; i < count; i++) {
+    msg = &m[i];
+    if (!jlog_file_pread(ctx->data, &msg->aligned_header, hdr_size, data_off_iter)) {
+      SYS_FAIL(JLOG_ERR_IDX_READ);
+    }
+    msg->header = &msg->aligned_header;
+    total_size += msg->header->mlen;
+    data_off_iter += (hdr_size + msg->header->mlen);
+  }
```

Comment:
We should make sure that the total_size isn't larger than the file size.

Reply:
Added bunches of checks for this.